src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -93,6 +108,173 @@ internal ModelPath WorksharingModelPath
+            }
+            return usedMaterialIds;
+        }
+
+        private List<Autodesk.Revit.DB.ElementId> PurgeMaterialAssets(List<Autodesk.Revit.DB.ElementId> elementIds)
+        {
+            List<ElementId> appearanceAssetIds = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(AppearanceAssetElement))
+                .ToElementIds()
+                .ToList();
+
+            List<ElementId> propertySet = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(PropertySetElement))
+                .ToElementIds()
+                .ToList();
+
+            int elementIdsCount = elementIds.Count();
+            for (int i = 0; i < elementIdsCount; i++)
+            {
+                try
+                {
+                    var material = this.InternalDocument.GetElement(elementIds[i]) as Autodesk.Revit.DB.Material;
+                    propertySet.Remove(material.ThermalAssetId);
+                    propertySet.Remove(material.StructuralAssetId);
+                    appearanceAssetIds.Remove(material.AppearanceAssetId);
+                }
+                catch (Exception)
+                {
+                    continue;
+                }
+            }
+
+            if (appearanceAssetIds.Count > 0)
+                this.InternalDocument.Delete(appearanceAssetIds);
+            if (propertySet.Count > 0)
+                this.InternalDocument.Delete(propertySet);
+
+            var purgedElementIds = new List<ElementId>(appearanceAssetIds.Concat(propertySet));
+            return purgedElementIds;
+        }
+
+        private List<ElementId> PurgeElements(bool runRecursively)
+        {
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            List<ElementId> purgeableElementIds = GetPurgeableElementIds();
+            
+            if (!runRecursively)
+            {
+                if (purgeableElementIds == null)
+                    return purgedElementIds;
+
+                purgedElementIds.AddRange(purgeableElementIds);
+                this.InternalDocument.Delete(purgeableElementIds);
+                return purgedElementIds;
+            }
+
+            if (purgeableElementIds == null)
+                return purgedElementIds;
+
+            purgedElementIds.AddRange(purgeableElementIds);
+            this.InternalDocument.Delete(purgeableElementIds);
+            purgedElementIds.AddRange(PurgeElements(runRecursively));
+
+            return purgedElementIds;
+        }
+
+        private List<Autodesk.Revit.DB.ElementId> GetPurgeableElementIds()
+        {
+            List<Autodesk.Revit.DB.FailureMessage> failureMessages = PerformanceAdviser.GetPerformanceAdviser()
+                                                                                       .ExecuteRules(this.InternalDocument, PerformanceAdviserRuleIds)
+                                                                                       .ToList();
```

Comment:
funky text formatting going on here

Reply:
not sure what you mean 😬

Reply:
it looked all over the place on my screen but seems fine now 🤷‍♂